Slab2 model

alarshi opened the pull request:
This PR adds an initial composition model that outputs 1 at the locations of the slab2 model (https://www.science.org/doi/10.1126/science.aat4723) and 0 otherwise. 
The slab2 model is input as an ascii boundary datafile with columns representing depths and thickness at the lat/long points on the Earth spaced at 0.05 degrees. This results in an ascii file that is too big to be added with aspect, and therefore a jupyter notebook is added to format the input slab2 model for an aspect input file.
Below is the paraview output from aspect using higher resolution on the test file.  

![slab2](https://user-images.githubusercontent.com/19296817/213247557-9e426210-1a9a-453a-9113-2714056465d5.png)

* [x] I have read the guidelines in our [CONTRIBUTING.md](../blob/main/CONTRIBUTING.md) document.

* [x] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [x] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../blob/main/tests/) directory.
* [ ] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.

gassmoeller commented on source/initial_composition/slab_model.cc:
```diff
@@ -0,0 +1,123 @@
+        {
+          const double slab_center    = slab_depths + slab_thickness/2.;
+          const double half_thickness = slab_thickness/2;
+          slab_composition            = ( 1 - std::tanh( 10 * (std::abs (slab_center - depth) - half_thickness)/half_thickness ) )/2;
```
At the moment you have hard-coded the thickness of the transition region and let it scale with the thickness of the slab. Is that  intended? I would have thought it is more useful to either have a constant transition region that is an input parameter (with the disadvantage like we discussed that it is only diffused in depth, not laterally). The alternative would be to simply not have a transition region at all, which may be simpler. E.g. something like this:
```suggestion
          slab_composition = 1.0;
```

gassmoeller commented on source/initial_composition/slab_model.cc:
```diff
@@ -0,0 +1,123 @@
+      const unsigned int slab_index          = this->introspection().compositional_index_for_name("slabs");
+
+      const double depth                     = this->get_geometry_model().depth(position);
+      double slab_composition                = 0;
```
```suggestion
      double slab_composition                = 0.0;
```

gassmoeller commented on source/initial_composition/slab_model.cc:
```diff
@@ -0,0 +1,123 @@
+  {
+    ASPECT_REGISTER_INITIAL_COMPOSITION_MODEL(SlabModel,
+                                              "slab model",
+                                              "An initial composition model computed to define slabs "
```
```suggestion
                                              "An initial composition model that defines slabs "
```

gassmoeller commented on source/initial_composition/slab_model.cc:
```diff
@@ -0,0 +1,123 @@
+                                              "an ascii format. The file describes the depths to the "
+                                              "top of the slabs and their thickness."
+                                              "The computed compositional value is 1 within the slabs "
+                                              "and zero elsewhere. In order to prevent sharp jumps, the "
+                                              "slabs are smoothened in the depth direction. "
```
Either adjust or remove.

```suggestion
                                              "and zero elsewhere. "
```

gassmoeller commented on source/initial_composition/slab_model.cc:
```diff
@@ -0,0 +1,123 @@
+                                              "Flamme, H., Furtney, M., & Smoczyk, G. M. (2018). Slab2, "
+                                              "a comprehensive subduction zone geometry model. Science, "
+                                              "362(6410), 58-61. The script to convert the slab2 model "
+                                              "into an aspect input data file is available.")
```
```suggestion
                                              "into an aspect input data file is available "
                                              "in the directory data/initial_composition/slab/.")
```

gassmoeller replied:
I think it should go into the data folder. We have some other python scripts already and they are mostly in the folders where they are needed. I see `contrib/python` more as a general folder for python scripts we do not know where else to put.

gassmoeller commented on tests/slab2_initial_composition_3d_shell.prm:
```diff
@@ -0,0 +1,110 @@
+  # datafile, generated using the jupyter notebook in
+  # (contrib/python) folder in the aspect base directory.
+  subsection Slab model
+    set Data directory       = /home/postdoc/Documents/
+    set Data file name       = slab2_depth_thickness.txt
```
We will have to find a way to make this test run on our tester machine. Can you include a very low resolution version of the slab files in `data/initial_composition/slab`? If that does not work, you could also make up a file with a new dataset (e.g. a single slab with a very large thickness). The important part is just that it can be included in the repository and can be used in this test.

gassmoeller commented on contrib/python/create_slab2_database.ipynb:
```diff
@@ -0,0 +1,267 @@
+{
```
Can you set this up so that the script automatically places the output data file in `data/initial_composition/slab`? You could also move the script over there.

gassmoeller reviewed (commented):
Thank you, this is great. 
I had a first look at the plugin and left a number of small comments, I did not look at the python notebook yet.
I think the main point to improve is to turn the test into something that can be run automatically, I left comments, but we can talk about it tomorrow as well.

alarshi commented on include/aspect/initial_composition/slab_model.h:
```diff
@@ -0,0 +1,92 @@
+*/
+
+
+#ifndef _aspect_initial_composition_slab2_model_h
+#define _aspect_initial_composition_slab2_model_h
+
```
Will do, thanks.

gassmoeller commented on source/initial_composition/slab_model.cc:
```diff
@@ -0,0 +1,123 @@
+    SlabModel<dim>::initialize ()
+    {
+      // The input slabs are defined from the surface of the model
+      std::set<types::boundary_id> surface_boundary_set;
+      surface_boundary_set.insert(this->get_geometry_model().translate_symbolic_boundary_name_to_id("top"));
+
+      // The two columns correspond to slabs depth and thickness
+      slab_boundary.initialize(surface_boundary_set, 2);
```
Please make the same change here that I did in #5028 for the other places where this operation happens.

alarshi commented:
Thank you, Rene, for the reviews! I added a low resolution data file in `initial-composition/slab` folder, that is now used for the test case in the recent commits. I also modified the notebook so that it outputs the data file in `initial-composition/slab` folder.

gassmoeller commented:
/rebuild

gassmoeller commented:
I think this is ready now. @jdannberg or someone else: could you do a review?

jdannberg commented on data/initial-composition/slab-model/create_slab2_database.py:
```diff
@@ -0,0 +1,164 @@
+    print (input_filenames[i])
+    slab_depths, slab_thickness = slab2_into_asciiboundary (input_filenames[i])
+
+# Save the arrays in the format for aspect ascii boundary :
+# 1. use radians in longitudes and latitudes, 2. first longitudes increase then latitudes
+# 3. convert depth and thickness from km to meters 4. depth is negative in the downward
+# direction of the slab2 database, convert to positive values. 5. Add necessary header
+# information.
```
Above it says the output is in colatitude, here it says latitude. Can you fix whichever is not correct?

jdannberg commented on source/initial_composition/slab_model.cc:
```diff
@@ -0,0 +1,139 @@
+    SlabModel<dim>::initialize ()
+    {
+      AssertThrow(this->introspection().compositional_name_exists("slabs"),
+                  ExcMessage("The initial composition plugin `slabs' did not find a "
```
```suggestion
                  ExcMessage("The initial composition plugin `slab model' did not find a "
```

jdannberg commented on source/initial_composition/slab_model.cc:
```diff
@@ -0,0 +1,139 @@
+      if (compositional_index != slab_index)
+        return 0.0;
+
+      // The first data column correspond to the slab depth and the second to slab thickness
```
```suggestion
      // The first data column corresponds to the slab depth and the second column to the slab thickness.
```

jdannberg commented on source/initial_composition/slab_model.cc:
```diff
@@ -0,0 +1,139 @@
+
+      // The first data column correspond to the slab depth and the second to slab thickness
+      // 'Slab depth' stands for the depth of the upper surface of the slab, 'Slab thickness'
+      // for the vertical distance between upper and lower surface. In lateral positions that
+      // do not have a slab thickness should be set to 0.0 and depth to a very large value
+      // that is outside the model domain.
```
I don't really understand that comment (also, I first read it as "positions that do not have a slab thickness"). 
Does it mean: In locations without slabs, we set the slab thickness to zero and the depth to a very large value? Or that in locations without slabs, the data file contains a zero slab thickness and a large depth? I guess the second?

jdannberg commented on source/initial_composition/slab_model.cc:
```diff
@@ -0,0 +1,139 @@
+                                              "into an aspect input data file is available in the directory "
+                                              "data/initial-composition/slab-model/. Please note that "
+                                              "Slab2 and the example data file assume spherical geometry "
+                                              "(latitude, longitude coordinates), however that is not "
```
```suggestion
                                              "(latitude, longitude coordinates), however, that is not "
```

jdannberg commented on tests/slab2_initial_composition_3d_shell.prm:
```diff
@@ -0,0 +1,91 @@
+
+subsection Initial composition model
+  set Model name = slab model
+  # The following lines needs to be modified such that
+  # they point to the location of the output slab2 model
```
```suggestion
  # The following lines point to the location of the output slab2 model
```

jdannberg commented on tests/slab2_initial_composition_3d_shell.prm:
```diff
@@ -0,0 +1,91 @@
+  # The following lines needs to be modified such that
+  # they point to the location of the output slab2 model
+  # datafile, generated using the jupyter notebook in
+  # (contrib/python) folder in the aspect base directory.
```
```suggestion
  # (/data/initial-composition/slab-model/) folder in the aspect base directory.
```

jdannberg commented on tests/slab2_initial_composition_3d_shell.prm:
```diff
@@ -0,0 +1,91 @@
+  set Initial global refinement                = 2
+  set Initial adaptive refinement              = 0
+  set Time steps between mesh refinement       = 0
+  set Strategy                                 = composition threshold, temperature
+  subsection Composition threshold
+    set Compositional field thresholds = 0.2
+  end
```
If there's no refinement, you don't really need that. 
```suggestion
```

jdannberg reviewed (approved):
I have some minor comments about the documentation, otherwise this looks good to me. Please also add a changelog entry.

jdannberg commented on data/initial-composition/slab-model/create_slab2_database.py:
```diff
@@ -0,0 +1,164 @@
+# coding: utf-8
+
+# This script uses the slab2 model data, available to download from
+# https://www.sciencebase.gov/catalog/item/5aa1b00ee4b0b1c392e86467, and
```
I went to that website and there are different things that can be downloaded. It would be good to either provide the exact download link, or specify what needs to be downloaded.

gassmoeller commented:
> I went to that website and there are different things that can be downloaded. It would be good to either provide the exact download link, or specify what needs to be downloaded.

Unfortunately the website does not allow to copy link locations, but I specified the exact name and heading from where to download the data.

I addressed all comments.

alarshi commented:
Thank you, Juliane, for the reviews! I pushed the changes, i.e., thickness is perpendicular to the slab surface instead of in the vertical direction, in the last commit .

gassmoeller commented:
@alarshi can you check the files in the data directory? I renamed the directory (to the same name as the plugin) and replaced the python notebook with a simpler python script. Can you integrate your changes into mine?

alarshi commented:
@gassmoeller : Sorry, I didn't mean to add the `slab` folder, must have happened during the merge. Does it look okay now? Also, do you want me to squash some commits?

gassmoeller commented:
Yes, can you please squash the commits? I usually do not care about that, but here we have several merge commits that can look quite confusing in a timeline of all commits.

alarshi commented:
@gassmoeller : I squashed the commits in the latest push.

gassmoeller reviewed (approved):
Looks all good to me. We should let @jdannberg have another look, but from my side this is ready.

jdannberg commented on data/initial-composition/slab-model/create_slab2_database.py:
```diff
@@ -0,0 +1,178 @@
+nan_indx = 0 
+def slab2_into_asciiboundary (slab_filename):
+    """
+    This function uses the above target mesh and fills it with the input slab depths and thicknesses. 
```
You also need to update the documentation here: 
```suggestion
    This function uses the above target mesh and fills it with the input slab depths, thicknesses and dip angles. 
```

jdannberg commented on source/initial_composition/slab_model.cc:
```diff
@@ -0,0 +1,137 @@
+                                              "into an aspect input data file is available in the directory "
+                                              "data/initial-composition/slab-model/. Please note that "
+                                              "Slab2 and the example data file assume spherical geometry "
+                                              "(latitude, longitude coordinates), however, that is not "
+                                              "necessary for this plugin, data files in "
+                                              "cartesian coordinates will work with box geometries.")
```
But in the initialization, we have an assert and the model will crash in cartesian coordinates. So at the moment, this does not work in box geometries.

alarshi replied:
I don't think the assert on the geometry model was intended to be there, I will remove it. Thanks!

jdannberg commented on source/initial_composition/slab_model.cc:
```diff
@@ -0,0 +1,137 @@
+    initial_composition (const Point<dim> &position,
+                         const unsigned int compositional_index) const
+    {
+      if (compositional_index != slab_index)
```
Sorry if this is somehow obvious, but: does this actually work? I don't see where the slab index is being set. 
In a previous version, there was a line
`const unsigned int slab_index          = this->introspection().compositional_index_for_name("slabs");`
and there was also an assert to make sure the slab index exists.

jdannberg reviewed (commented):
I just have two comments about documentation, but feel free to address them in a follow-up PR. 

But I am confused about the slab index. If that's just a misunderstanding on my part, feel free to merge.

alarshi commented:
@jdannberg : Thank you for reviewing it! I addressed your comments in the latest push. I hope I did not break anything from the previous changes :(

gassmoeller reviewed (approved):
Thank you, this looks ready to merge now. I added a changelog entry and pushed the commit, because for some reason the testers did not trigger yesterday (looks like a problem on github's side).

Ready to merge if the testers are happy.
